### ISSUE TITLE
Fetch deeper history in repo?

### ISSUE THREAD

**Issue by max-sixty** [GoogleCloudPlatform/cloud-builders#43]
We need to pull a deeper history for our repos (because they use an automatic versioning scheme that requires knowing how far `HEAD` is from the prior tag)

This has generally worked:
```
- name: gcr.io/cloud-builders/git
  args: [fetch, --depth=100]
```

But with one repo I get this error:
```
BUILD
Step #0: Already have image (with digest): gcr.io/cloud-builders/git
Starting Step #0
Step #0: fatal: missing blob object 'a702af496a55a81c491fe2eef993939a26c0d8a6'
Step #0: error: https://source.developers.google.com/p/$project/r/$repo did not send all necessary objects
Step #0: 
Finished Step #0
```

Any thoughts on the cause?

**Comment by MaximilianR** [comment]
I believe I may have a corrupted repo in GH. I will confirm and close if so

**Comment by MaximilianR** [comment]
This does seem to be a problem specifically with a) the commands the Builder uses to initiate the repo and b) Google Source Repository.

After running the commands locally listed [here](https://github.com/GoogleCloudPlatform/cloud-builders/issues/26), running `git fetch --unshallow`:

- Works fine if `$REMOTE` is the GitHub repo
- Breaks with the commands above if `$REMOTE` is the Google hosted repo

This seems to only occur if there are many tags - when I remove all but the latest tag, this works fine.

But at the moment Google Source Repository + Google Cloud Builder isn't usable for us

**Comment by MaximilianR** [comment]
FYI this is still a breaking issue - latest build-id if anyone is on these: `1be7d2d4-651c-44d6-8d06-d1dabb782514`

**Comment by ImJasonH** [comment]
Thanks for reporting this.

By any chance, does `git fetch --unshallow` work?

**Comment by MaximilianR** [comment]
`--unshallow` does seem to work! Thank you

Closing this for now

**Comment by tsloughter** [comment]
Is this thread saying that:

```
- name: gcr.io/cloud-builders/git
  args: [fetch, --unshallow]
```

Should work? Is it supposed to work with github triggered builds? I am still getting "did not send all necessary objects" with that.

Our release versions are based on the latest tag + commit info and the git info a build currently has is not enough for it to pick up the latest tag, so like `git describe --tags --long` fails.

**Comment by tsloughter** [comment]
And while `args: [fetch]` does work to fetch all the branches. The `HEAD` must still detached because describe continues to fail.

**Comment by MaximilianR** [comment]
It worked in one instance

Do you want to open a new issue given you can repro? Or maintainers: feel free to open this for tracking

**Comment by tsloughter** [comment]
I figured I'd check if i understood this issue before opening a new one, though it is the same issue I believe, so maybe reopening makes more sense?

**Comment by tsloughter** [comment]
It is extra weird since it worked once. I created a new trigger on a branch `cloudbuild-test` and added a step after the `fetch` that does `['checkout', 'develop']` and that build was able to build the version based on the latest tag, like a `git describe --tags --long`. But when I do the same checkout from a trigger on the `develop` branch it is again detached.

**Comment by MaximilianR** [comment]
One thing to check is whether it works on GitHub based repos. You can repro the steps with the link above

**Comment by tsloughter** [comment]
Do you know what it will set to `$REV`? Does it use the branch name or does github send the latest commit ref from the branch in the trigger?

**Comment by MaximilianR** [comment]
> Do you know what it will set to $REV? Does it use the branch name or does github send the latest commit ref from the branch in the trigger?

Fair question; I'm not sure

**Comment by antoniomo** [comment]
Just bumping this, as nothing seems to work for me. Our use case is similar, on our build pipeline we would like to get some git history/reflog to see what needs rebuilding from a monorepo with several services, but even after `--unshallow`, `git reflog show $BRANCH_NAME --pretty='%h'` or `git reflog show origin/$BRANCH_NAME --pretty='%h'` both show only the latest commit.

**Comment by bendory** [comment]
This GitHub issue tracker is intended for bugs with the officially supported builder images specifically.
Please report issues and feature requests regarding the GCB service to our public issue tracker at
https://issuetracker.google.com/issues/new?component=190802&template=1162743

**Comment by jli** [comment]
When I tried adding the `fetch` step for my Github repo, I got this error from Cloud Build: `fatal: could not read Username for 'https://github.com': No such device or address`

I was able to get it to work following these docs: ["Accessing GitHub from a build via SSH keys "](https://cloud.google.com/build/docs/access-github-from-build) (linked from ["Including the repository history in a build"](https://cloud.google.com/build/docs/automating-builds/create-manage-triggers#including_the_repository_history_in_a_build)).

Steps from the docs that I followed exactly:
- enabled Secret Manager
- generated a new ssh keypair: added private key as a Secret Manager secret, added public key to my Github profile
- granted Secret Manager access to the Cloud Build service account
- added step to `cloudbuild.yaml` config for setting up ssh private key and `known_hosts` file

I deviated from the docs for the other part of the ["Configure the build"](https://cloud.google.com/build/docs/access-github-from-build#configure_the_build) section, though. The steps in the docs are for making a new clone of the repo, but I just wanted to extend the history of the shallow clone I already had. So, instead of the `git clone` step, I added this step:

```yaml
- id: 'fetch more history'
  name: gcr.io/cloud-builders/git
  entrypoint: bash
  args:
    - -c
    - |
      git remote set-url origin 'git@github.com:<GITHUB_USERNAME>/<GITHUB_REPO>'
      git fetch --depth=300
  volumes:
  - name: 'ssh'
    path: /root/.ssh
```

Sorry for bumping this old issue, but hope this helps someone who stumbled upon this issue while searching like I did!
